reasoner/planner/ConjunctionStreamPlan.java

```diff
@@ -0,0 +1,337 @@
+            // If we did everything right, we should have an easy build.
+            if (subConjunction.size() == 1) {
+                VariableSets variableSets = VariableSets.determineVariableSets(list(), subConjunction, availableInputs, requiredOutputs);
+                //  use resolvableOutputs instead of rightOutputs because this node has to do the job of the parent as well - joining the identifiers
+                Set<Retrievable> resolvableOutputs = VariableSets.difference(requiredOutputs, variableSets.extendOutputWith);
+                return new ResolvablePlan(subConjunction.get(0), variableSets.rightInputs, variableSets.extendOutputWith, resolvableOutputs);
+            } else {
+                List<Resolvable<?>> suffix = subConjunction.subList(1, subConjunction.size());
+                VariableSets variableSets = VariableSets.determineVariableSets(subConjunction.subList(0, 1), subConjunction.subList(1, subConjunction.size()), availableInputs, requiredOutputs);
+                ConjunctionStreamPlan leftPlan = new ResolvablePlan(subConjunction.get(0), variableSets.leftIdentifiers, set(), variableSets.leftOutputs);
+                ConjunctionStreamPlan rightPlan = buildRecursiveSuffix(suffix, variableSets.rightInputs, variableSets.rightOutputs);
+                return new CompoundStreamPlan(list(leftPlan, rightPlan),
+                        variableSets.identifiers, variableSets.extendOutputWith, requiredOutputs);
+            }
+        }
+
+        public Pair<List<Resolvable<?>>, List<Resolvable<?>>> divide(List<Resolvable<?>> subConjunction) {
+            int splitAfter = findSplitAfterIndex(subConjunction);
+            List<Resolvable<?>> left = new ArrayList<>();
+            List<Resolvable<?>> right = new ArrayList<>();
+            for (int j = 0; j < subConjunction.size(); j++) {
+                if (j <= splitAfter) left.add(subConjunction.get(j));
+                else right.add(subConjunction.get(j));
+            }
+
+            return new Pair<>(left, right);
+        }
+
+        private int findSplitAfterIndex(List<Resolvable<?>> subConjunction) {
+            Set<Retrievable> nextSuffixVariables = new HashSet<>(subConjunction.get(subConjunction.size() - 1).retrieves());
+            int r;
+            for (r = subConjunction.size() - 2; r > 0; r--) { // r > 0 because we need atleast one thing in our left
+                nextSuffixVariables.addAll(subConjunction.get(r).retrieves());
+                Set<Retrievable> subtreeBounds = VariableSets.intersection(boundsAt.get(r), nextSuffixVariables);
+                Set<Retrievable> candidateFirstChildBounds = VariableSets.intersection(subConjunction.get(r).retrieves(), boundsAt.get(r));
+
+                if (!VariableSets.difference(subtreeBounds, candidateFirstChildBounds).isEmpty()) {
+                    break;
+                }
+            }
+            return r;
+        }
```

Comment:
(Is supposed to) split the (sub)conjunction into two so that the right-tree satisfies the properties described in the PR.